### PR TITLE
[Fix] fa3_mla: dequantize only referenced FP8 KV pages

### LINKED ISSUE #35291: [Bug] MLA + fp8 KV: fa3 and flashinfer_mla dequantise the ENTIRE per-layer KV pool every layer, every forward (43.7% of a decode step on DeepSeek-R1 / 8xH200)
## Summary

While benchmarking a MegaMoE (NVFP4) MoE backend integration on DeepSeek-R1, we profiled the
decode path and found that the single most expensive GPU kernel had nothing to do with MoE.

On both the `fa3` and `flashinfer_mla` backends, the absorbed-MLA path materialises a bf16 copy
of the **entire per-layer KV pool** on every layer of every forward pass:

```python
kv_cache = self.token_to_kv_pool.get_key_buffer(layer.layer_id).to(q.dtype)
```

`get_key_buffer` returns the whole per-layer pool, so with an fp8 KV cache this is a full-pool
`fp8_e4m3 -> bf16` dequantisation costing **O(max_total_tokens) per layer per forward** —
independent of batch size, sequence length, and of how much of the pool is actually live.

Measured on DeepSeek-R1-0528 (NVIDIA ModelOpt NVFP4 checkpoint, fp8 KV cache), 8×H200,
`--max-total-tokens 131072`: one 225.79 µs kernel firing **61× per decode step** =
**13.77 ms per decode step**, with **zero overlap** with any other GPU work — 43.7% of a
31.52 ms step. About 98.9% of the bytes moved are never referenced by the page table.

Switching to `--attention-backend flashmla --page-size 64` removes it entirely while keeping
fp8 KV at full capacity: decode step **31.52 ms → 17.68 ms (1.78×)**, GSM8K accuracy unchanged.

## Affected configurations

All three conditions must hold at once:

1. **fp8 KV cache.** In our case it comes from the checkpoint, not the command line:
   `hf_quant_config.json: {"quant_algo": "NVFP4", "kv_cache_quant_algo": "FP8"}`.
   `server_args` still prints `kv_cache_dtype='auto'`.
2. **MLA / absorbed attention**, i.e. `head_dim = kv_lora_rank 512 + qk_rope 64 = 576 > 256`
   (DeepSeek family).
3. **`fa3` or `flashinfer_mla`.** With `--attention-backend` unspecified our server selected
   fa3: `Attention backend not specified. Use fa3 backend by default.`

`flashmla`, `trtllm_mla` and `triton` call `get_key_buffer(layer.layer_id)` with **no** `.to()`
and are unaffected.

**Note this is not a config error on our side.** `mem_cache/kv_cache_dtype.py` correctly resolves
`auto` + `kv_cache_quant_algo == "FP8"` into `kv_cache_dtype_str = "fp8_e4m3"`. The fp8-native
branch is skipped **only** because of the head_dim gate:

```python
if (
    self.kv_cache_dtype_str != "auto"   # passes
    and layer.head_dim <= 256           # FAILS for MLA (576)
    and not self.kv_cache_is_mxfp8
):
    ...
    q = q.to(self.kv_cache_dtype)       # the cheap direction: cast the tiny Q instead
```

FA3's own non-MLA path casts the small Q **down** rather than casting the large pool **up**.

## Minimal reproduction — single GPU, no model, no serving stack

```python
import torch
# 576 = kv_lora_rank 512 + qk_rope_head_dim 64 ; +1 = the null/sentinel page
src = torch.empty(131073, 576, dtype=torch.float8_e4m3fn, device="cuda")
dst = src.to(torch.bfloat16)      # this is the kernel, once per layer per forward
```

Measured on H200 (torch 2.13.0+cu130):

```
max_total_tokens= 65536  N=37,749,312  grid.x= 73730   113.4 us/call  1.00 TB/s
max_total_tokens=131072  N=75,498,048  grid.x=147458   222.2 us/call  1.02 TB/s
```

`grid.x = 147458` matches the in-server trace exactly, and 222.2 µs vs the in-server 225.79 µs
(1.6% apart). The 65536 case is exactly half, confirming the cost is `O(max_total_tokens)`.

**Full repro:** serve any MLA model whose KV pool resolves to fp8 with the default backend, and
profile steady-state decode. Look for one `unrolled_elementwise_kernel` carrying
`LoadWithCast`/`StoreWithCast`, firing once per layer, with
`grid.x = ceil((max_total_tokens + 1) * 576 / 512)`. Its duration is **invariant to batch size** —
that invariance is the signature.

## Where it is in the source

Verified present in `main` on **2026-08-18** by direct fetch of the raw files —
**four sites, two backends**, all under the comment `# Do absorbed multi-latent attention`,
and **none of them guarded** by any dtype check, helper wrapper, or try/except:

| File | Function |
|---|---|
| `python/sglang/srt/layers/attention/flashattention_backend.py` | `forward_extend` |
| `python/sglang/srt/layers/attention/flashattention_backend.py` | `forward_decode` |
| `python/sglang/srt/layers/attention/flashinfer_mla_backend.py` | `forward_extend` |
| `python/sglang/srt/layers/attention/flashinfer_mla_backend.py` | `forward_decode` |

## Measured impact

Torch profiler, GPU activities, 5 steady-state decode steps, concurrency 8:

| metric | value |
|---|---|
| calls per 5 decode steps | 305 = **61 layers × 5** (fires in the 3 dense layers too) |
| duration | 225.79 µs (pstdev 0.41; min 225.02, max 228.22 — one tight cluster) |
| cost per decode step | **13.77 ms (43.7% of a 31.52 ms step)** |
| overlap with any other kernel | **0 of 305**; nearest other kernel 51.55 µs away |
| live KV rows at the profiled instant | ~1,440 of 131,073 ⇒ **~98.9% of the traffic is dead** |

Attribution (no python stack was available — the kernels are replayed via `cudaGraphLaunch`):

<details>
<summary>launch geometry, byte cross-check, temporal position</summary>

- `launch_unrolled_kernel` uses 128 threads × 4 elements ⇒ 512/block,
  `grid.x = ceil(N/512) = 147458` ⇒ `N ∈ (75,497,984 , 75,498,496]`.
  `(131072 + 1) × 576 = 75,498,048` — inside, and tight to one row
  (`131072 × 576` would give grid 147456, not 147458).
- `131073 × 576 × 1 B × 61 layers = 4.289 GiB` = the logged
  `KV Cache is allocated. dtype: torch.float8_e4m3fn, #tokens: 131072, KV size: 4.29 GB`.
- `LoadWithCast<1>` / `StoreWithCast<1>` ⇒ a dtype-converting copy;
  72 MiB read + 144 MiB written in 225.79 µs = 1.00 TB/s.
- Temporal position, 305/305 with no exception:
  `set_mla_kv_buffer` → **[this copy]** → `flash::prepare_varlen_num_blocks` → FlashAttn.

</details>

## Control experiment

Same server, `--kv-cache-dtype bfloat16`, `max_total_tokens` pinned to 65536 for both arms
(both logged `dtype: torch.bfloat16, #tokens: 65536, KV size: 4.29 GB`):

| MoE backend | KV dtype | pool | big copies | decode step |
|---|---|---:|---:|---:|
| A | fp8 | 131072 | 305 | 31.52 ms |
| B | fp8 | 131072 | 305 | 37.00 ms |
| A | bf16 | 65536 | **0** | **17.20 ms** |
| B | bf16 | 65536 | **0** | **22.86 ms** |

The kernel **vanishes** (0, not smaller), on two entirely different MoE backends whose own
totals differ by only 0.08% — confirming this cost is independent of the MoE path.

bf16 KV is only a control: it halves KV capacity at equal memory.

## Workaround (what we ship)

`--attention-backend flashmla --page-size 64` — keeps fp8 KV at **full 131072** capacity and
produces zero such launches:

| | fa3 (default) | flashmla |
|---|---:|---:|
| decode step | 31.52 ms | **17.68 ms** |
| GSM8K 5-shot, 200 q, T=0 | 0.960 | **0.965** (invalid 0.000 both) |

Caveat: flashmla loses fa3's fused `set_mla_kv_buffer_kernel` and uses eager
`cat`/`index_put`/`quant` instead, costing +0.425 ms/step. Net gain still 13.35 ms/step.

## Why this is easy to miss

- It does not crash; it only costs time.
- `kv_cache_dtype` prints as `auto`, so nothing in the launch line hints that the pool is fp8.
- It needs fp8 KV **and** MLA **and** one of the two affected backends, simultaneously.
- fa3 is what the server selects when `--attention-backend` is unspecified.

## Prior art

- **PR #34638** "[MLA] Gather only the batch's KV rows for fp8 paged prefill" describes the
  *same idiom* in its own words ("up-converts the whole per-layer KV pool", "O(pool) per layer
  per forward", 17.76 GiB/layer → ~29 MB/layer) and fixes it via
  `token_to_kv_pool.get_mla_kv_buffer(...)`. It targets `flashinfer_mla_backend.py`
  `forward_extend` only and states **"Decode is not addressed."**
  As of 2026-08-18 its change is **not in `main`** — we fetched the file and both `.to(q.dtype)`
  sites are still present, and `get_mla_kv_buffer` is not called there.
- **PR #7148** "Fix FP8 KV Cache Support in FA3 Backend" (merged 2025-06-25) introduced both
  `.to(q.dtype)` sites and the `head_dim <= 256` gate, with no benchmarks. Two users in that
  thread reported DeepSeek/MLA + fp8 KV being no faster or slower than bf16; the response
  attributed it to in-kernel descale overhead, which does not apply to the MLA path because it
  never reaches the descale branch. **The symptom was observed at the time but never localised.**
We also scanned the last 25 commits touching `flashattention_backend.py` (through 2026-08-18):
none address fp8 KV, the KV pool, dequantisation, or absorbed-MLA performance.
If we have missed an existing PR or issue, please point us at it and close this as a duplicate.

- **Issue #30815** "FP8 KV-cache decode slowdown" (fa3, decode, open) is a **different** defect:
  write-side unfused quantisation in `set_kv_buffer` plus a per-layer Q cast, ~0.4–1 ms/step.

## Possible directions

1. Let the MLA kernel consume fp8 directly (what `flashmla` / `trtllm_mla` / `triton` do).
2. Gather and cast only the page-table-referenced rows, as PR #34638 does for prefill.
   For **decode** this additionally needs a page-table remap and must stay CUDA-graph-safe —
   all our 305 launches carry `graph id 23`, i.e. they run inside a captured decode graph.
   This is presumably why #34638 stops at prefill.
3. At minimum, warn when `attention_backend in {fa3, flashinfer_mla}` and the resolved pool dtype
   is fp8 and `head_dim > 256`. Detection should use the **pool dtype**
   (`token_to_kv_pool.dtype`), never the CLI string.

## What we did not do

- We have **not** implemented or tested a fix.
- We profiled **decode only**; the `forward_extend` sites were read, not measured.
- We did not verify how vLLM or TensorRT-LLM handle this.
- The "43.7% of a step" figure is specific to `max_total_tokens=131072`; the cost is proportional
  to it.

## Environment

| | |
|---|---|
| model | DeepSeek-R1-0528, NVIDIA ModelOpt NVFP4 checkpoint |
| hardware | 8× H200 SXM |
| parallelism | TP = DP = EP = 8, `--enable-dp-attention` |
| flags | `--quantization modelopt_fp4 --mem-fraction-static 0.56 --max-total-tokens 131072`, page_size 1, **no** `--attention-backend`, **no** `--kv-cache-dtype` |
| KV cache (log) | `dtype: torch.float8_e4m3fn, #tokens: 131072, KV size: 4.29 GB` |
| image | `lmsysorg/sglang:v0.5.x` (digest `sha256:4a5e481e…`) |

### PR DESCRIPTION
**Pull request by ggboooy**
## Motivation and scope

Related to #35291. This PR addresses the FA3 MLA path and does not automatically close the issue. The corresponding native FP8 solution for `flashinfer_mla` is tracked separately in #35322, which covers decode and paged prefill on qualifying SM90 configurations and is still open.

FA3 absorbed MLA currently converts the entire per-layer FP8 KV pool to the query dtype on every forward, even when the page table references only a small fraction of the pool. This change converts only the referenced physical pages.

## Modifications

- Add a Triton kernel for referenced-page FP8 E4M3/E5M2 to BF16/FP16 conversion, preserving physical page IDs and the original page table.
- Support multi-token pages: compute live pages with `ceil(cache_seqlen / page_size)` and map physical page IDs to token rows. Copy shared tail pages in full regardless of which request claims the page first.
- Share a persistent shadow across MLA layers on a rank. Persistent epoch state deduplicates pages and supports CUDA Graph capture/replay.
- Enable the optimized path for normal FA3 MLA decode and eager extend; real FA3 comparisons cover page sizes 1, 16, 64, and 128.
- Preserve fallback behavior for speculative, cascade, local/SWA, cross-attention, CP/DCP, draft, unified-dense, and other unsupported configurations.
- Synchronize with current main and use the upstream `cp_active` state for the CP fallback guard.

## Correctness tests

Revalidated on the final integrated tree on 2026-09-21: H800/SM90, PyTorch 2.9.1+cu129.

```text
70 passed, 48 subtests passed
0 failed, 0 skipped
```

Coverage includes:

- Backend/source eligibility and every disqualifier, including fallback guards with larger page sizes.
- Exact E4M3/E5M2 to BF16/FP16 conversion; page sizes 1/3/16/64/128; page boundaries, shared tails, highest physical page, empty sequences, and invalid inputs.
- Cross-layer workspace reuse, page recycling, and CUDA Graph replay with updated source data, page tables, and sequence lengths.
- Padded, non-contiguous page-table rows; out-of-range sentinels in inactive entries and empty requests; bounded-program traversal across multiple pages.
- Bitwise equality against full-pool conversion through real FA3 with page sizes 1/16/64/128 and query lengths 1/8 (`rtol=0`, `atol=0`).

Ruff lint/format, isort, codespell, Python compilation, test-registration checks, registered-test placement checks, pytest entry-point checks, and `git diff --check` pass locally. GitHub Lint passed on the identical code tree before the history-only squash to a single commit. The squashed commit is `c78a96e97725d0d462da215c3afe1d236c314732`. The full upstream test workflow is gated on the maintainer-controlled `run-ci` label; its tests have not run.

## End-to-end performance and output comparison

These measurements were collected on 2026-09-17, before synchronizing with current main. Performance has not been rerun on the newly integrated main revision.

Single H800/SM90, real DeepSeek-V2-Lite-Chat weights, TP=1, FA3, FP8 E4M3 KV, 131072-token pool, decode CUDA Graph enabled, radix cache disabled. The baseline uses the same source tree with shadow allocation disabled to select the original full-pool conversion. Each workload was warmed up and measured five times; values below are medians.

| Page size | Concurrency | Input/output tokens | Full-pool conversion | Selective conversion | Output throughput: baseline → selective |
|---:|---:|---:|---:|---:|---:|
| 1 | 8 | 1024/256 | 2.8342 s | 1.3878 s | 722.6 → 1475.7 token/s |
| 64 | 8 | 1024/256 | 2.8388 s | 1.3619 s | 721.4 → 1503.7 token/s |

Across two page sizes, five workloads, and five measured repetitions per version, 1140 output sequences containing 166400 generated tokens matched token-for-token within each page-size/workload comparison. Counts include repeated workloads for consistency validation; this is not a task-accuracy evaluation.

In a short decode profile, 108 full-pool conversions took approximately 24.84 ms in total. The selective kernels took approximately 0.50 ms at page size 1 and 1.15 ms at page size 64, with no corresponding full-pool conversion kernel in the optimized traces. A CUDA Graph microbenchmark with 131073 pool rows, width 576, and 1440 live tokens reduced page-size-1 conversion from 225.82 μs to 5.60 μs. This kernel-level speedup is not an end-to-end speedup.

Both temporary end-to-end test trees used the same `only_qv=False` adapter for the test host's older FA3 interface. That environment adapter is not part of this PR. Results should not be extrapolated directly to multi-GPU DeepSeek-R1, other live-KV ratios, or production traffic distributions.

## Checklist

- [x] Formatting and static checks.
- [x] Registered CPU/GPU regression tests.
- [x] Correctness, end-to-end performance, and profiling results.
- [x] Explicit FA3/FlashInfer scope without auto-closing #35291.
- [ ] Documentation update (not applicable: no new public API or configuration).

---
### CI States

Latest PR Test (Base): :x: [Run #35599611565](https://github.com/sgl-project/sglang/actions/runs/35599611565)
Latest PR Test (Extra): :x: [Run #35599611156](https://github.com/sgl-project/sglang/actions/runs/35599611156)
Latest PR Test (AMD ROCm 10): :x: [Run #35599611586](https://github.com/sgl-project/sglang/actions/runs/35599611586)

### CODE REVIEW

**Comment by janbernloehr** [comment]
Thanks for addressing the FA3 path. One concern: this appears to be a partial fix for #35291 rather than a complete one.
The issue identifies four full-pool casts:
- `flashattention_backend.py`: `forward_extend` and `forward_decode`
- `flashinfer_mla_backend.py`: `forward_extend` and `forward_decode`

This PR replaces the two `flashattention_backend.py` sites, but leaves both `flashinfer_mla_backend.py` sites unchanged. Those paths therefore still perform the original `get_key_buffer(layer.layer_id).to(q.dtype)` operation over the entire KV pool. PR #34638 addresses FlashInfer-MLA forward_extend, but explicitly does not cover decode.
In addition, the new FA3 shadow is enabled only for `page_size == 1`; larger page sizes intentionally fall back to the original full-pool conversion.
Could we either cover the remaining paths or change `Fixes #35291` to indicate that this fixes only the FA3/page-size-1 subset, so #35291 is not automatically closed while the FlashInfer-MLA and larger-page cases remain?

**Comment by ggboooy** [comment]
Thanks for pointing out the scope issue. The FlashInfer MLA side is already covered by the separate proposal in #35322: it wires up FlashInfer's native FP8 MLA path for qualifying SM90 configurations, covering decode and paged prefill without the full-pool BF16 cast. That PR is still open and has not been merged.

This PR focuses on the FA3 MLA backend. I have now extended the selective conversion to multi-token pages and added regression coverage for page boundaries, shared tails, page reuse, CUDA Graph replay, real FA3 equivalence, padded page tables, and bounded-program traversal. After synchronizing with current main and adapting the CP fallback guard to `cp_active`, the final integrated tree passes **70 tests and 48 subtests on SM90, with no failures or skips**.

I also changed the description from `Fixes #35291` to a scoped reference so this PR will not automatically close the broader issue while other backend work remains.

The history is now squashed to a single commit with only the four intended files changed. GitHub Lint passed on the identical code tree before this history-only squash. The full upstream test workflow is currently waiting for the maintainer-controlled `run-ci` label; those tests have not run yet.
